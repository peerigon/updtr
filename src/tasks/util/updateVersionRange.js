import semver from "semver";

// Matches a semver version range that can be transformed to the new version in a safe manner
const expectedSemverPattern = /^(\^|~|>=|)(\d+)\.(\d+|x|\*)\.(\d+|x|\*)(-[a-z][a-z\-.\d]+|)$/i;
const numberPattern = /^\d+$/;

function parse(semverString) {
    const match = semverString.match(expectedSemverPattern);

    return match === null ?
        null :
    {
        operator: match[1],
        major: match[2],
        minor: match[3],
        patch: match[4],
        release: match[5],
    };
}

function isPinned({ operator, major, minor, patch }) {
    return operator === "" &&
        [major, minor, patch].every(
            version => numberPattern.test(version) === true
        );
}

function tryVersionRangeUpdate(parsedOldRange, parsedNewVersion) {
    const { minor, patch, operator } = parsedOldRange;
    let newMinor = parsedNewVersion.minor;
    let newPatch = parsedNewVersion.patch;

    if (numberPattern.test(minor) === false) {
        newMinor = minor;
        newPatch = numberPattern.test(patch) === true ? minor : patch;
    } else if (numberPattern.test(patch) === false) {
        newPatch = patch;
    }

    return operator +
        parsedNewVersion.major +
        "." +
        newMinor +
        "." +
        newPatch +
        parsedNewVersion.release;
}

function isExpectedNewVersion(parsedNewVersion) {
    return parsedNewVersion !== null && parsedNewVersion.operator === "";
}

function fallbackRange(newVersion) {
    return "^" + newVersion;
}

/**
 * Tries to apply the newVersion while maintaining the range (see https://github.com/peerigon/updtr/issues/47)
 * This is kind of risky because there are tons of semver possibilities. That's why this
 * function is very conservative in accepting semver ranges. If the range is not easily updatable,
 * we opt-out to npm's default caret operator.
 *
 * @param {string} oldRange
 * @param {string} newVersion
 * @returns {string}
 */
export default function updateVersionRange(oldRange, newVersion) {
    const parsedOldRange = parse(oldRange.trim());

    if (parsedOldRange !== null) {
        if (isPinned(parsedOldRange) === true) {
            // The old version was pinned, so the new should also be pinned
            return newVersion;
        }

        const parsedNewVersion = parse(newVersion);

        if (isExpectedNewVersion(parsedNewVersion) === false) {
            return newVersion;
        }

        const newVersionRange = tryVersionRangeUpdate(
            parsedOldRange,
            parsedNewVersion
        );

        // All this is kind of error prone so let's do a sanity check if everything's ok
        if (semver.satisfies(newVersion, newVersionRange) === true) {
            return newVersionRange;
        }
    }

    return fallbackRange(newVersion);
}
