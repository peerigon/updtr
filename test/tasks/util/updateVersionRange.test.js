import updateVersionRange from "../../../src/tasks/util/updateVersionRange";
import versionRanges from "../../fixtures/versionRanges";

describe("updateVersionRange()", () => {
    it("should match the expected new version range", () => {
        const newVersionRanges = versionRanges.map(oldVersionRange =>
            updateVersionRange(oldVersionRange, "5.0.0"));
        const comparisons = versionRanges.map(
            (oldVersionRange, index) =>
                oldVersionRange + "     becomes     " + newVersionRanges[index]
        );

        expect(comparisons).toMatchSnapshot();
    });
    describe("when the new version contains a release tag", () => {
        it("should match the expected new version range", () => {
            const newVersionRanges = versionRanges.map(oldVersionRange =>
                updateVersionRange(oldVersionRange, "5.0.0-beta.1"));
            const comparisons = versionRanges.map(
                (oldVersionRange, index) =>
                    oldVersionRange +
                    "     becomes     " +
                    newVersionRanges[index]
            );

            expect(comparisons).toMatchSnapshot();
        });
    });
});
