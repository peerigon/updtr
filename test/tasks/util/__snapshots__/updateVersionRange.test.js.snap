// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateVersionRange() should match the expected new version range 1`] = `
Array [
  "~0.6.1-1     becomes     ^5.0.0",
  "1.0.0 - 2.0.0     becomes     ^5.0.0",
  "1.0.0-2.0.0     becomes     ^5.0.0",
  "1.0.0     becomes     5.0.0",
  "   1.0.0        becomes     5.0.0",
  "1.0.0-beta.1     becomes     5.0.0",
  ">=1.0.0-beta.1     becomes     >=5.0.0",
  "1.0.0-BETA.1     becomes     5.0.0",
  "1.0.0-alpha-1     becomes     5.0.0",
  "1.0.0 alpha     becomes     ^5.0.0",
  "~1.0.0     becomes     ~5.0.0",
  "   ~1.0.0        becomes     ~5.0.0",
  "~1.0.x     becomes     ~5.0.x",
  ">=*     becomes     ^5.0.0",
  "     becomes     ^5.0.0",
  "*     becomes     ^5.0.0",
  ">=1.0.0     becomes     >=5.0.0",
  "   >=1.0.0        becomes     >=5.0.0",
  ">1.0.0     becomes     ^5.0.0",
  "   >1.0.0         becomes     ^5.0.0",
  "<=2.0.0     becomes     ^5.0.0",
  "<2.0.0     becomes     ^5.0.0",
  ">=   1.0.0     becomes     ^5.0.0",
  "> 1.0.0     becomes     ^5.0.0",
  "<=   2.0.0     becomes     ^5.0.0",
  "<    2.0.0     becomes     ^5.0.0",
  "<	2.0.0     becomes     ^5.0.0",
  ">=0.1.97     becomes     >=5.0.0",
  "0.1.20 || 1.2.4     becomes     ^5.0.0",
  ">=0.2.3 || <0.0.1     becomes     ^5.0.0",
  "||     becomes     ^5.0.0",
  "2.x.x     becomes     5.x.x",
  "1.2.x     becomes     5.0.x",
  "1.2.x || 2.x     becomes     ^5.0.0",
  "x     becomes     ^5.0.0",
  "2.*.*     becomes     5.*.*",
  "1.2.*     becomes     5.0.*",
  "1.2.* || 2.*     becomes     ^5.0.0",
  "2     becomes     ^5.0.0",
  "2.3     becomes     ^5.0.0",
  "~2.4     becomes     ^5.0.0",
  "~>3.2.1     becomes     ^5.0.0",
  "~1     becomes     ^5.0.0",
  "~>1     becomes     ^5.0.0",
  "~> 1     becomes     ^5.0.0",
  "~1.0     becomes     ^5.0.0",
  "~ 1.0     becomes     ^5.0.0",
  ">=1     becomes     ^5.0.0",
  ">= 1     becomes     ^5.0.0",
  "<1.2     becomes     ^5.0.0",
  "< 1.2     becomes     ^5.0.0",
  "1     becomes     ^5.0.0",
  "~v0.5.4-pre     becomes     ^5.0.0",
  "~v0.5.4-pre     becomes     ^5.0.0",
  "=0.7.x     becomes     ^5.0.0",
  ">=0.7.x     becomes     >=5.0.x",
  ">=0.7.X     becomes     >=5.0.X",
  "=0.7.x     becomes     ^5.0.0",
  ">=0.7.x     becomes     >=5.0.x",
  "<=0.7.x     becomes     ^5.0.0",
  ">0.2.3 >0.2.4 <=0.2.5     becomes     ^5.0.0",
  ">=0.2.3 <=0.2.4     becomes     ^5.0.0",
  "1.0.0 - 2.0.0     becomes     ^5.0.0",
  "^1     becomes     ^5.0.0",
  "^3.0.0     becomes     ^5.0.0",
  "^1.0.0 || ~2.0.1     becomes     ^5.0.0",
  "^0.1.0 || ~3.0.1 || 5.0.0     becomes     ^5.0.0",
  "^0.1.0 || ~3.0.1 || >4 <=5.0.0     becomes     ^5.0.0",
  "*.1.2     becomes     ^5.0.0",
  "x.1.2     becomes     ^5.0.0",
  "1.x.2     becomes     5.x.x",
  "1.*.2     becomes     5.*.*",
]
`;

exports[`updateVersionRange() when the new version contains a release tag should match the expected new version range 1`] = `
Array [
  "~0.6.1-1     becomes     ^5.0.0-beta.1",
  "1.0.0 - 2.0.0     becomes     ^5.0.0-beta.1",
  "1.0.0-2.0.0     becomes     ^5.0.0-beta.1",
  "1.0.0     becomes     5.0.0-beta.1",
  "   1.0.0        becomes     5.0.0-beta.1",
  "1.0.0-beta.1     becomes     5.0.0-beta.1",
  ">=1.0.0-beta.1     becomes     >=5.0.0-beta.1",
  "1.0.0-BETA.1     becomes     5.0.0-beta.1",
  "1.0.0-alpha-1     becomes     5.0.0-beta.1",
  "1.0.0 alpha     becomes     ^5.0.0-beta.1",
  "~1.0.0     becomes     ~5.0.0-beta.1",
  "   ~1.0.0        becomes     ~5.0.0-beta.1",
  "~1.0.x     becomes     ^5.0.0-beta.1",
  ">=*     becomes     ^5.0.0-beta.1",
  "     becomes     ^5.0.0-beta.1",
  "*     becomes     ^5.0.0-beta.1",
  ">=1.0.0     becomes     >=5.0.0-beta.1",
  "   >=1.0.0        becomes     >=5.0.0-beta.1",
  ">1.0.0     becomes     ^5.0.0-beta.1",
  "   >1.0.0         becomes     ^5.0.0-beta.1",
  "<=2.0.0     becomes     ^5.0.0-beta.1",
  "<2.0.0     becomes     ^5.0.0-beta.1",
  ">=   1.0.0     becomes     ^5.0.0-beta.1",
  "> 1.0.0     becomes     ^5.0.0-beta.1",
  "<=   2.0.0     becomes     ^5.0.0-beta.1",
  "<    2.0.0     becomes     ^5.0.0-beta.1",
  "<	2.0.0     becomes     ^5.0.0-beta.1",
  ">=0.1.97     becomes     >=5.0.0-beta.1",
  "0.1.20 || 1.2.4     becomes     ^5.0.0-beta.1",
  ">=0.2.3 || <0.0.1     becomes     ^5.0.0-beta.1",
  "||     becomes     ^5.0.0-beta.1",
  "2.x.x     becomes     ^5.0.0-beta.1",
  "1.2.x     becomes     ^5.0.0-beta.1",
  "1.2.x || 2.x     becomes     ^5.0.0-beta.1",
  "x     becomes     ^5.0.0-beta.1",
  "2.*.*     becomes     ^5.0.0-beta.1",
  "1.2.*     becomes     ^5.0.0-beta.1",
  "1.2.* || 2.*     becomes     ^5.0.0-beta.1",
  "2     becomes     ^5.0.0-beta.1",
  "2.3     becomes     ^5.0.0-beta.1",
  "~2.4     becomes     ^5.0.0-beta.1",
  "~>3.2.1     becomes     ^5.0.0-beta.1",
  "~1     becomes     ^5.0.0-beta.1",
  "~>1     becomes     ^5.0.0-beta.1",
  "~> 1     becomes     ^5.0.0-beta.1",
  "~1.0     becomes     ^5.0.0-beta.1",
  "~ 1.0     becomes     ^5.0.0-beta.1",
  ">=1     becomes     ^5.0.0-beta.1",
  ">= 1     becomes     ^5.0.0-beta.1",
  "<1.2     becomes     ^5.0.0-beta.1",
  "< 1.2     becomes     ^5.0.0-beta.1",
  "1     becomes     ^5.0.0-beta.1",
  "~v0.5.4-pre     becomes     ^5.0.0-beta.1",
  "~v0.5.4-pre     becomes     ^5.0.0-beta.1",
  "=0.7.x     becomes     ^5.0.0-beta.1",
  ">=0.7.x     becomes     ^5.0.0-beta.1",
  ">=0.7.X     becomes     ^5.0.0-beta.1",
  "=0.7.x     becomes     ^5.0.0-beta.1",
  ">=0.7.x     becomes     ^5.0.0-beta.1",
  "<=0.7.x     becomes     ^5.0.0-beta.1",
  ">0.2.3 >0.2.4 <=0.2.5     becomes     ^5.0.0-beta.1",
  ">=0.2.3 <=0.2.4     becomes     ^5.0.0-beta.1",
  "1.0.0 - 2.0.0     becomes     ^5.0.0-beta.1",
  "^1     becomes     ^5.0.0-beta.1",
  "^3.0.0     becomes     ^5.0.0-beta.1",
  "^1.0.0 || ~2.0.1     becomes     ^5.0.0-beta.1",
  "^0.1.0 || ~3.0.1 || 5.0.0     becomes     ^5.0.0-beta.1",
  "^0.1.0 || ~3.0.1 || >4 <=5.0.0     becomes     ^5.0.0-beta.1",
  "*.1.2     becomes     ^5.0.0-beta.1",
  "x.1.2     becomes     ^5.0.0-beta.1",
  "1.x.2     becomes     ^5.0.0-beta.1",
  "1.*.2     becomes     ^5.0.0-beta.1",
]
`;
